index.rst

```diff
@@ -10,20 +10,357 @@
+While that's arbitrarily long in general, I don't think it's unreasonable to either tell users that refs will get stale after a while, or just add timestamps to signed URIs so we can spot them and refresh them as necessary when ``get`` is called.
+
+Butler methods vs. Butler.registry methods
+------------------------------------------
+
+One outcome of :jira:`RFC-888` was that users disliked having to remember which aspects of the butler public interface were on the `Butler` class vs. the `Registry` it holds.
+It's also confusing that `Butler.registry` and `Butler.datastore` both appear to be public attributes, but only the former really is (and some of its methods are not really intended for external use, either).
+Moving all of the public `Registry` interface to `Butler` and making `Butler.registry` (and `Butler.datastore`) private would be a major change, but it's the kind of change that would also help us with other changes:
+
+- It lets us repurpose `Registry` as an internal polymorphic interface focused on abstracting over the differences between a direct SQL backend and an http backend, while leaving common user-focused client code to `Butler`.
+
+- It gives us a clear boundary and deprecation point for other needed (or at least desirable) API changes, in that new versions of methods can differ from their current ones without having to work out a deprecation path that allows new and old behavior to be coexist in the same signature.
+
+In addition to moving convenience code out of `Registry` and into `Butler`, we'll also need to move our caching (of collections, dataset types, and certain dimension records) to the client, and it'll certainly be better to put that in one client class (i.e. `Butler`) that replicate it across both `Registry` client implementations.
+
+At some point, we may opt to continue backwards compatibility support for `Butler.registry` methods by making `Butler.registry` return a lightweight proxy that forwards back to `Butler` instead of a real `Registry` instance.
+
+Batch operations and unifying bulk transfers
+--------------------------------------------
+
+The current Butler provides a `transaction` method that returns a context manager that attempts to guarantee consistency for all operations performed in that context.
+This only works rigorously for Registry operations at present, as the Datastore transaction system we have at present is not fault tolerant.
+Much of the rest of this proposal is motivated by trying to address this, but our current transaction interface is not really viable even for Registry-only operations when an http client/server implementation is required.
+
+Instead, the prototype includes a `RawBatch` class that represents a low-level serializable batch of multiple `Registry` operations to be performed together within one transaction, and a `Butler.batched` method and `BatchHelper` class to provide a high-level interface for constructing them.
```

Comment:
Having client-side code prepare RawBatches could lead to extended locking of the database.  Can we mitigate by somehow limiting the size of RawBatches?

Reply:
Shouldn't be hard at all - we can pretty easily get a rough estimate of the number of rows modified by a raw batch, and just put a cap on that in both the client (for convenience) and server (for rigor).